resources: updated pannotia documentation for compile commands

KaiBatley opened the pull request:
The docker command for compiling pannotia tests needs to have visibility to m5ops.

jebraun3 committed:
Correcting FW kernel launches

Change-Id: I8352cce84c098b17e566915c215248860b9a599d

jebraun3 committed:
Correcting FW result verification

Change-Id: Ia59c67ff83f2b24047d0f1e51b99b6a2a700ed9a

jebraun3 committed:
Adding mmap support to FW

Change-Id: I07df1ee28925bc46cf7c3f74189ce41403b3801c

ivanaamit committed:
resources: Update references with new URL for Docker images (#14)

Updated all references to reference the new URL for GitHub Docker
Registry instead of Google Registry.
Issue: [#336](https://github.com/gem5/gem5/issues/336)

mattsinc commented on src/gpu/pannotia/bc/README.md:
```diff
@@ -15,11 +15,10 @@ Compiling BC, compiling the GCN3_X86/Vega_X86 versions of gem5, and running BC o
 ## Compilation and Running
 
-To compile BC:
+To compile BC assuming gem5 and gem5-resources are in your working directory:
```
"assuming the gem5 and gem5-resources repositories"

mattsinc commented on src/gpu/pannotia/color/README.md:
```diff
@@ -15,20 +15,18 @@ Compiling both CLR variants, compiling the GCN3_X86/Vega_X86 versions of gem5, a
 ## Compilation and Running
 
-To compile Color:
+To compile Color assuming gem5 and gem5-resources are in your working directory:
```
same here

mattsinc commented on src/gpu/pannotia/fw/README.md:
```diff
@@ -15,11 +15,10 @@ Compiling FW, compiling the GCN3_X86/Vega_X86 versions of gem5, and running FW o
 ## Compilation and Running
 
-To compile FW:
+To compile FW assuming gem5 and gem5-resources are in your working directory:
```
same here

mattsinc commented on src/gpu/pannotia/bc/README.md:
````diff
@@ -15,11 +15,10 @@ Compiling BC, compiling the GCN3_X86/Vega_X86 versions of gem5, and running BC o
 ```
-cd src/gpu/pannotia/bc
-docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID ghcr.io/gem5/gcn-gpu make gem5-fusion
+docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID ghcr.io/gem5/gcn-gpu bash -c "cd gem5-resources/src/gpu/pannotia/bc ; make gem5-fusion"
````
probably we should explain why we cd instead of making the working directory different (because of m5ops, I assume)?

mattsinc commented on src/gpu/pannotia/mis/README.md:
```diff
@@ -15,11 +15,10 @@ Compiling MIS, compiling the GCN3_X86/Vega_X86 versions of gem5, and running MIS
 ## Compilation and Running
 
-To compile MIS:
+To compile MIS assuming gem5 and gem5-resources are in your working directory:
```
same here

mattsinc commented on src/gpu/pannotia/pagerank/README.md:
```diff
@@ -15,18 +15,16 @@ Compiling both PageRank variants, compiling the GCN3_X86/Vega_X86 versions of ge
 ## Compilation and Running
 
-PR has two variants: default and spmv.  To compile the "default" variant:
+PR has two variants: default and spmv.  To compile the "default" variant assuming gem5 and gem5-resources are in your working directory:
```
same here

mattsinc commented on src/gpu/pannotia/sssp/README.md:
```diff
@@ -15,18 +15,16 @@ Compiling both SSSP variants, compiling the GCN3_X86/Vega_X86 versions of gem5,
 ## Compilation and Running
 
-SSSP has two variants: csr and ell.  To compile the "csr" variant:
+SSSP has two variants: csr and ell.  To compile the "csr" variant assuming gem5 and gem5-resources are in your working directory:
```
same here

mattsinc reviewed (changes requested):
In addition to the minor changes, I think we should probably also update the documentation to explain that the m5ops are required, and how to compile them?

KaiBatley commented:
@mattsinc I've changed the wording and added the explanation that m5ops are required with the command to compile them. In addition, I moved the GCN3_X86/gem5.opt compile info before the test compile info as it's needed to clone gem5 before compiling m5ops. As compiling and running the test both assume gem5 and gem5-resources are in the same directory I updated the Makefiles so that the gem5 path does not have to be specified.

mattsinc commented on src/gpu/pannotia/bc/Makefile.gem5-fusion:
```diff
@@ -8,7 +8,7 @@ HIP_PATH ?= /opt/rocm/hip
 
 # these are needed for m5ops
-GEM5_PATH ?= /path/to/gem5
+GEM5_PATH ?= $(CURDIR)/../../../../../gem5
```
CURDIR does not seem any better.  And the levels of "../" here are unclear -- where does that come from?  Probably it would be better to set it to $(GEM5_ROOT) and tell users this variable must be set?

KaiBatley replied:
The goal is to take the CURDIR of ~/gem5-resources/src/gpu/pannotia/<test> and get to ~/gem5. The ../.. would exit gem5-resources and go into gem5 without the user having to specify anything. If the layout of repositories was different the user should still be able to specify GEM5_PATH=/path/to/gem5 on the command line. Setting $(GEM5_ROOT) would do the same thing correct or am I missing something?

mattsinc replied:
In my opinion, the problem with CURDIR is it is not intuitive.  GEM5_ROOT (I think) has a clear meaning.

mattsinc commented on src/gpu/pannotia/bc/README.md:
```diff
@@ -13,17 +13,6 @@ Betweenness Centrality (BC) is a graph analytics application that is part of the
-
-If you use the Makefile.default file instead, the Makefile will generate code designed to run on the real GPU instead.  Moreover, note that Makefile.gem5-fusion requires you to set the GEM5_ROOT variable (either on the command line or by modifying the Makefile), because the Pannotia applications have been updated to use [m5ops](https://www.gem5.org/documentation/general_docs/m5ops/).  By default, ...
-
 ## Compiling GCN3_X86/gem5.opt
```
Probably this title needs to be change since now it is an introduction instead of directions on compilation?

KaiBatley replied:
Added a new title to introduce compilation to match the later title for running the test.

mattsinc commented on src/gpu/pannotia/bc/README.md:
````diff
@@ -34,6 +23,22 @@ cd gem5
+docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID ghcr.io/gem5/gcn-gpu bash -c "cd gem5/util/m5 ; scons build/x86/out/m5"
+```
+
+To compile BC assuming gem5 and gem5-resources are in your working directory:
````
Still need to update this to say "... assuming the gem5 and gem5-resources repositories ..."

mattsinc replied:
Also I don't think the updated text explains why we cd?

KaiBatley replied:
I missed updating the text for bc and one other test, fixed now. I've also added a sentence explaining the cd before the m5 and test commands.

mattsinc reviewed (changes requested):
I went through BC in detail, but it looks like the changes for it also need to be applied to the others.  This is getting closer though, thanks for your hard work on this.

mattsinc commented on src/gpu/pannotia/fw/README.md:
```diff
@@ -34,27 +25,44 @@ cd gem5
+# Running FW on GCN3_X86/gem5.opt
 
-# Assuming gem5 and gem5-resources are in your working directory
+## Assuming gem5 and gem5-resources are in your working directory
```
Should there be text between this and 49?  Else it seems like a title with no content.

mattsinc commented on src/gpu/pannotia/pagerank/README.md:
```diff
@@ -13,47 +13,54 @@ PageRank (PR) is a graph analytics application that is part of the Pannotia benc
 Compiling both PageRank variants, compiling the GCN3_X86/Vega_X86 versions of gem5, and running both PageRank variants on gem5 is dependent on the gcn-gpu docker image, `util/dockerfiles/gcn-gpu/Dockerfile` on the [gem5 stable branch](https://gem5.googlesource.com/public/gem5/+/refs/heads/stable).
 
-## Compilation and Running
+# Compiling BC and GCN3_X86/gem5.opt
```
Title is wrong -- should be PR, not BC

mattsinc commented on src/gpu/pannotia/sssp/README.md:
```diff
@@ -13,47 +13,54 @@ Single-Source Shortest Path (sssp) is a graph analytics application that is part
 Compiling both SSSP variants, compiling the GCN3_X86/Vega_X86 versions of gem5, and running both SSSP variants on gem5 is dependent on the gcn-gpu docker image, `util/dockerfiles/gcn-gpu/Dockerfile` on the [gem5 stable branch](https://gem5.googlesource.com/public/gem5/+/refs/heads/stable).
 
-## Compilation and Running
+# Compiling BC and GCN3_X86/gem5.opt
```
Title is wrong -- should be SSSP, not BC

mattsinc reviewed (changes requested):
I believe all apps other than BC have the same title problem, but I did not mark all of them.

The CURDIR thing I continue to thing is probably not going to be obvious to others.

KaiBatley commented:
Fixed title naming of tests, fixed the fw test having multiple titles and added GEM5_ROOT to makefile.

mattsinc commented on src/gpu/pannotia/color/README.md:
```diff
@@ -13,49 +13,58 @@ Graph Coloring (CLR) is a graph analytics application that is part of the Pannot
+  
 
-If you use the Makefile.default file instead, the Makefile will generate code designed to run on the real GPU instead.  Moreover, note that Makefile.gem5-fusion requires you to set the GEM5_ROOT variable (either on the command line or by modifying the Makefile), because the Pannotia applications have been updated to use [m5ops](https://www.gem5.org/documentation/general_docs/m5ops/).  By default, ...
+To compile Color assuming the gem5 and gem5-resources repositories are in your working directory:
```
Are words missing here?  Because it ends on ":", but then the next line appears to be unrelated.

mattsinc commented on src/gpu/pannotia/fw/README.md:
````diff
@@ -34,27 +25,46 @@ cd gem5
+docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID ghcr.io/gem5/gcn-gpu bash -c "cd gem5/util/m5 ; scons build/x86/out/m5"
+```
+Note that Makefile.gem5-fusion requires you to set the GEM5_ROOT variable (either on the command line or by modifying the Makefile)  
+  
````
Extra line here, remove

mattsinc commented on src/gpu/pannotia/fw/README.md:
````diff
@@ -34,27 +25,46 @@ cd gem5
+```
+
+If you use the Makefile.default file instead, the Makefile will generate code designed to run on the real GPU instead. By default, the Makefile builds for gfx801 and gfx803, and is placed in the src/gpu/pannotia/fw/bin folder. FW can be run on a non-mmapped input file, used to generate an mmapped input file, or run on an mmapped input file. To run FW using an mmapped input file, you must generate ...
+
````
Extra line, remove

mattsinc commented on src/gpu/pannotia/mis/Makefile.gem5-fusion:
```diff
@@ -6,7 +6,8 @@ HIPCC = $(HIP_PATH)/bin/hipcc
 # these are needed for m5ops
 # TODO: Need some sort of explicit PATH?  Read in?
-GEM5_PATH ?= /nobackup/sinclair/gem5
+GEM5_ROOT ?= /nobackup/sinclair/gem5
```
Probably we should make this /path/to/gem5 like the others?

mattsinc commented on src/gpu/pannotia/fw/Makefile.gem5-fusion:
```diff
@@ -3,7 +3,8 @@ HIPCC = $(HIP_PATH)/bin/hipcc
 # these are needed for m5ops
 # TODO: Need some sort of explicit PATH?  Read in?
-GEM5_PATH ?= /nobackup/sinclair/gem5
+GEM5_ROOT ?= /nobackup/sinclair/gem5
```
Probably we should make this /path/to/gem5 like the others?

mattsinc commented on src/gpu/pannotia/pagerank/Makefile.gem5-fusion:
```diff
@@ -4,7 +4,8 @@ OPTS = -O3
 # these are needed for m5ops
 # TODO: Need some sort of explicit PATH?  Read in?
-GEM5_PATH ?= /nobackup/sinclair/gem5
+GEM5_ROOT ?= /nobackup/sinclair/gem5
```
Probably we should make this /path/to/gem5 like the others?

mattsinc commented on src/gpu/pannotia/sssp/Makefile.gem5-fusion:
```diff
@@ -3,7 +3,8 @@ HIPCC = $(HIP_PATH)/bin/hipcc
 # these are needed for m5ops
 # TODO: Need some sort of explicit PATH?  Read in?
-GEM5_PATH ?= /nobackup/sinclair/gem5
+GEM5_ROOT ?= /nobackup/sinclair/gem5
```
Probably we should make this /path/to/gem5 like the others?

mattsinc reviewed (changes requested):
A few minor updates, but almost there!

KaiBatley commented:
Updated path to /path/to/gem5. Also fixed extra lines and changed ':' to make it more clear.

mattsinc commented on src/gpu/pannotia/bc/README.md:
```diff
@@ -34,9 +25,26 @@ cd gem5
+
+If you use the Makefile.default file instead, the Makefile will generate code designed to run on the real GPU instead. By default, the Makefile builds for gfx801 and gfx803, and is placed in the src/gpu/pannotia/bc/bin folder.
+
+
```
extra line here, please remove

mattsinc commented on src/gpu/pannotia/bc/Makefile.gem5-fusion:
```diff
@@ -8,7 +8,8 @@ HIP_PATH ?= /opt/rocm/hip
 
 # these are needed for m5ops
-GEM5_PATH ?= /path/to/gem5
+GEM5_ROOT ?= /path/to/gem5
```
as I reflect on this change, I wonder if it would be better to remove GEM5_PATH completely and just use GEM5_ROOT?  Right now I'm not sure having both adds much value ... but I've already made you jump through a lot of hoops here so it's probably fine to let it be.

KaiBatley replied:
The only reason I didn't do that was if GEM5_PATH was being used it would break things. The weekly.sh tests for example do "export GEM5_PATH=${gem5_root} ; make gem5-fusion" (although weekly.sh is now deprecated?). If it's not used elsewhere (outside automated testing?) I can update the weekly.sh and makefile to GEM5_ROOT.

mattsinc replied:
Fair enough.  Don't update weekly.sh though, I'd rather leave this as is.

@abmerop 's comments remain though.

BobbyRBruce replied:
> The only reason I didn't do that was if GEM5_PATH was being used it would break things. The weekly.sh tests for example do "export GEM5_PATH=${gem5_root} ; make gem5-fusion" (although weekly.sh is now deprecated?). If it's not used elsewhere (outside automated testing?) I can update the weekly.sh and makefile to GEM5_ROOT. 

Since the weekly tests are deprecated can you just use `GEM5_ROOT`? Having `GEM5_PATH` here is confusing.

mattsinc reviewed (changes requested):
At least one small change.

abmerop reviewed (changes requested):
This should be based on stable branch IMO.  It looks like all of the instructions apply to the current working version. Jason recommended using stable if that is the case.

Additionally, as is, this would clobber all of the changes I made to these Makefiles (e.g., adding gfx902 target).

mattsinc commented:
> This should be based on stable branch IMO. It looks like all of the instructions apply to the current working version. Jason recommended using stable if that is the case.
> 
> Additionally, as is, this would clobber all of the changes I made to these Makefiles (e.g., adding gfx902 target).

It sounds like @KaiBatley needs to rebase his develop to incorporate your change?  I'm surprised the checks wouldn't have caught that...

Even if Kai goes to stable, wouldn't we want your changes (already committed?) to be not clashing?

abmerop commented:
I am not sure how stable/develop are kept in sync. Perhaps a question for @BobbyRBruce or @powerjg

powerjg commented:
The rule is as follows:
If the change applies to the *current gem5 stable*, then the change should be on the *stable* branch of gem5-resources
If the change *cannot work* on gem5 stable and requires updates to gem5 that are only found on gem5 develop, then the change should be on the *develop* branch of gem5-resources

When gem5 is released, the stable and develop branches are merged together. When gem5-resources's stable and develop branches diverge, then we merge stable into develop.

Note that these are the same rules for the website repo as well.

@ivanaamit can you make sure that this information can be found in the CONTRIBUTING.md or README.md file of gem5-resources and gem5-website? Also, double check with Bobby that my description is precise.

ivanaamit commented:
@powerjg, I will do it.

BobbyRBruce commented:
> The rule is as follows: If the change applies to the _current gem5 stable_, then the change should be on the _stable_ branch of gem5-resources If the change _cannot work_ on gem5 stable and requires updates to gem5 that are only found on gem5 develop, then the change should be on the _develop_ branch of gem5-resources
> 
> When gem5 is released, the stable and develop branches are merged together. When gem5-resources's stable and develop branches diverge, then we merge stable into develop.
> 
> Note that these are the same rules for the website repo as well.
> 
> @ivanaamit can you make sure that this information can be found in the CONTRIBUTING.md or README.md file of gem5-resources and gem5-website? Also, double check with Bobby that my description is precise.

This is all correct. As pointed out to me recently, calling the "main" gem5-resources branch `stable` is confusing because it's not stable. Unlike the gem5 code repo, it's permitted to change at any time. In hindsight gem5 Resource's `stable` branch should be called something like `production`, `live`, or `release`.

For gem5 Resources, i periodically merge `stable` into `develop` and, as Jason mentioned, merged `develop` into `stable` upon a release. In addition to what Jason said, there's some additional notes on this process found in our release procedures documentation: https://www.gem5.org/documentation/general_docs/development/release_procedures/.

KaiBatley commented:
Pull #14 has updated the same pannotia test run commands on develop. If I were to switch to stable it would create merge conflicts I'm not sure would easily resolve and may lead to doubling work. If it's advised I still switch to stable I assume I'll just have to open a new pull request as I can't switch the source branch to stable?

powerjg commented:
Maybe we should first merge current develop into stable, then you can move this change to stable? Let us know if this is required/wanted.

As far as the merge conflicts go, *someone* is going to have to resolve them. It would be best for the contributor to resolve them since they have the knowledge needed.

KaiBatley commented:
Merging develop into stable would be preferred to avoid the merge conflicts.

powerjg commented:
@BobbyRBruce can you merge develop into stable and then @KaiBatley can rebase on top of stable.

BobbyRBruce commented:
> @BobbyRBruce can you merge develop into stable and then @KaiBatley can rebase on top of stable.

We do that on a release and we've not released v23.1 yet. 

I don't mean to come across as  snippy, but what is the matter with resolving these merges?


From my understanding, all you need to do is create a new branch from stable, cherry pick change on top of it, resolve the merge conflicts, force push to to this PR, then change merge target.

powerjg commented:
My understanding is that there's changes in develop that should have been on stable.

BobbyRBruce commented:
> My understanding is that there's changes in develop that should have been on stable.

Then, can't we just cherry pick those changes as well? Or just apply those changes manually to stable too?

powerjg commented:
Yeah, that makes sense! @KaiBatley can you give us the exact commit hashes and links in github for the changes that should have been on develop to make it easier for you to rebase this change?

KaiBatley commented:
This commit (c6a1ed0d9fd2f8b97e1dac97520d7cc04b6d82b8) on PR #14 should be on stable

mattsinc commented:
> This commit ([c6a1ed0](https://github.com/gem5/gem5-resources/commit/c6a1ed0d9fd2f8b97e1dac97520d7cc04b6d82b8)) on PR #14 should be on stable

I can help you resolve the merge conflicts if needed, @KaiBatley

mattsinc commented:
> Yeah, that makes sense! @KaiBatley can you give us the exact commit hashes and links in github for the changes that should have been on develop to make it easier for you to rebase this change?

@BobbyRBruce I am actually a bit confused here.  Why would we want to cherry pick commits from develop underneath?  Those commits will have different SHAs then, and maybe clash when you do merge develop with stable for this in a couple weeks -- at least that is how cherry-picking has always worked in my experience.

mattsinc commented:
> > The rule is as follows: If the change applies to the _current gem5 stable_, then the change should be on the _stable_ branch of gem5-resources If the change _cannot work_ on gem5 stable and requires updates to gem5 that are only found on gem5 develop, then the change should be on the _develop_ branch of gem5-resources
> > When gem5 is released, the stable and develop branches are merged together. When gem5-resources's stable and develop branches diverge, then we merge stable into develop.
> > Note that these are the same rules for the website repo as well.
> > @ivanaamit can you make sure that this information can be found in the CONTRIBUTING.md or README.md file of gem5-resources and gem5-website? Also, double check with Bobby that my description is precise.
> 
> This is all correct. As pointed out to me recently, calling the "main" gem5-resources branch `stable` is confusing because it's not stable. Unlike the gem5 code repo, it's permitted to change at any time. In hindsight gem5 Resource's `stable` branch should be called something like `production`, `live`, or `release`.
> 
> For gem5 Resources, i periodically merge `stable` into `develop` and, as Jason mentioned, merged `develop` into `stable` upon a release. In addition to what Jason said, there's some additional notes on this process found in our release procedures documentation: https://www.gem5.org/documentation/general_docs/development/release_procedures/.

I must admit I've read these two posts (and the linked webpage) at least 3 times now, and I'm still confused.  Ultimately my mental model has always been that we should be developing on the develop branch -- and only on stable if it's a hotfix.  Even if we changed the name to production, my mental model would stay this way.

It seems that is not the case from reading this, or at least is not the case for gem5-resources?  And so I thought the root issue was that @KaiBatley should have been making these changes on develop but didn't -- but the replies here seem to indicate that Kai made the changes where they were supposed to be, but @abmerop and James had done their changes in the wrong place previously?

Ultimately this probably has little to do with Kai's patch, but is ultimately highlighting something I clearly don't understand about our pull request process.

mattsinc commented:
Regardless of my questions above, @KaiBatley here are the steps you'd need to take (I think):

1.  Create a new branch that is based on gem5/stable’s current head (i.e., without your change).  Sidenote: this part may be slightly tricky if the branch name is different from KaiBatley:develop since that is what you named the branch you pushed this change on ... I am not sure what the best solution here is, but one solution would be to use git rebase -i HEAD~N to remove all patches currently on develop that are not on stable from your local KaiBatley:develop branch.
2.  Cherry pick https://github.com/gem5/gem5-resources/commit/c6a1ed0d9fd2f8b97e1dac97520d7cc04b6d82b8 onto your new branch.  I am not sure if the updated SHA will be a problem when @BobbyRBruce merges stable and develop, but I will defer to Bobby there.
3.  Cherry pick https://github.com/gem5/gem5-resources/commit/5187949395ba8cc05ad2dd1247881bdcc1f0f3d8 onto your new branch.  I am not sure if the updated SHA will be a problem when @BobbyRBruce merges stable and develop, but I will defer to Bobby there.
4.  Cherry pick your commit onto your new branch.  This will cause a merge conflict with steps 2 and 3.
5.  Resolve merge conflicts – make sure to retain the changes in steps 2 and 3, but in the format you have changed things to in step 4.
6.  Force push this update for your pull request.

mattsinc reviewed (dismissed):
As far as I can tell, this integrates @abmerop and @jebraun3's develop changes properly with yours (e.g., FW mmap text is there, gfx902 is there, ghcr is there) and puts all on stable, so I'm approving.  @abmerop , @BobbyRBruce , and @powerjg may want to ok that things are as they expect now though.

mattsinc commented:
And thanks/nice work @KaiBatley !

abmerop commented on src/gpu-fs/README.md:
```diff
@@ -77,9 +77,9 @@ You can obtain the `vmlinux-5.4.0-105-generic` kernel using the following path f
 It is sometimes useful to build your own application and run in gem5. A docker is provided to allow users to build applications without needing to install ROCm locally. A pre-built docker image is available on gcr.io. This image can be pulled then used to build as follows:
```
FYI I was planning to delete this file in #12 but this reminds me that I should probably move these instructions there on how to build applications for GPUFS.  Thank you for the updates all the same though!

mattsinc replied:
Perhaps inadvertently, but @abmerop shouldn't line 77 above be ghcr.io instead of gcr.io?  If so, that can probably be in a different patch though.

abmerop replied:
Yes probably, good catch.  I believe both *might* work but my understanding is gcr costs money for the Davis folks and ghcr is preferred

mattsinc replied:
Can you submit a separate pull request for this then?

abmerop replied:
I will put it in #12 if that's ok.  I need to update this anyway for https://github.com/gem5/gem5/pull/571

abmerop reviewed (dismissed):
Looks good to me, thanks for the updates!

mattsinc commented:
@BobbyRBruce / @powerjg : will wait for one/both of you to chime in before we merge this in.

BobbyRBruce commented:
> > Yeah, that makes sense! @KaiBatley can you give us the exact commit hashes and links in github for the changes that should have been on develop to make it easier for you to rebase this change?
> 
> @BobbyRBruce I am actually a bit confused here. Why would we want to cherry pick commits from develop underneath? Those commits will have different SHAs then, and maybe clash when you do merge develop with stable for this in a couple weeks -- at least that is how cherry-picking has always worked in my experience.

Cherry-picking does result in "duplicate" commits in this case, though it shouldn't cause any additional strain when merging. Git knows that if two diffs describe the same modification to the same line/codeblock then to create one diff from two you only need to have that change declared once and there's no conflict to resolve in this case because both are inagreement. It doesn't matter if they are different diffs in other respects. Merging only ever becomes difficult when diffs disagree on a change. Merging is just comparing diffs so the same change on two separate branches is fine

To get a feel of how git works imagine two branches, `X` and `Y`, both of which diverged from commit `A` to add a file to the repository. The file created on both branches is identical but branch `X` created it over 10 commits in a manner of iterative improvement with each commit amending to the file, fixing typos, or refactoring things until eventually reaching the desired state. On the other hand Branch `Y` added the file in one perfect commit. You can note here that there's no common hashes between the branches despite the fact they've implemented exactly the same thing

Now, if one were to be merged into the other git would do so by creating a unified diff of each branch that describes the diff from `A` to that branch's head. Branch `Y`'s unified diff is exactly the same as the change described in its singular, lone commit, while `X`'s would be a equivalent to the squashing of it's 10. Git would then compare the two unified diffs and see if there are are disagreements on how a a piece of of code should change. In any case where both diffs are trying to apply the same change to the same piece of code, git can say there's no conflict. In cases where one diff modifies one piece of code, and one modifies another, there's no conflict. It's only when the diffs try to modify the same code in different ways that git declares a "merge conflict" the user must must decide how these two changes work together. This resolution is what is stored in the merge commit. In our case the unified diff of `X` and `Y` are both identical so a merge commit is created which says nothing needs resolved in this case.

This differs from rebasing which is more error prone. If you tried to rebase or 10 commit `X` branch  on top of `Y` you'd likely face issues. If the first commit in `X` started off with a partial file full of errors, git will compare that to  Y's already completed solution and see many conflicts to serolve. Here git would ask you to, esentially, alter the commit to sit neatly atop branch `X`, with no conflicts. Then you'd do the same process for all the other commits in `X`, likely needing to do the same fixing procedure for the rest as the changes from X cascade down the chain of commits brought over from `Y`.


The real pain comes from when branches diverge too much, and that just happens in any system. This is why I periodically merge stable into develop so I can handle the conflicts in small chunks instead of having them explode on us during a release.

Anyway, in short, I have no problem with cherry-picking. I'll review this patch in full later today.

BobbyRBruce commented on src/gpu/pannotia/fw/Floyd-Warshall.cpp:
```diff
@@ -84,67 +91,150 @@
+	    tmpchar = optarg;
+            break;
+	case 'h':  // Help
+            fprintf(stderr, "SWITCHES\n    -f [file name]\n            input file name\n");
+            fprintf(stderr, "    -m [mode]\n            operation mode: default (run without mmap), generate, usemmap\n");
+	    fprintf(stderr, "    -v,    verify results\n");
```
This isn't a request for any changes but I think splitting strings like this over multiple lines would improve readability and take some of the pain out of figuring out these awkward spacings. In CPP any two strings separated by only whitespace or newline characters are parsed as a single string. So you can do stuff like this:


```cpp
fprintf(stderr, "SWITCHES\n"
                "    -f [file name]\n"
                "            input file **name\n""
                "    -m [mode]\n"
                "            operation mode: default (run without mmap),\n"
                "                            generate,\n"
                "                            usemmap\n"
                "    -v,     verify results\n");
```

Also, the last `fprintf` isn't intended the same as the rest.

mattsinc replied:
As @KaiBatley mentioned, this would be for @jebraun3 to fix, since it was inherited from the cherry-picked patches.  Do you want it as part of this patch, @BobbyRBruce ?

BobbyRBruce replied:
Marking as resolved. Not to do in this PR. Current solution is sufficient for now.

BobbyRBruce commented on src/gpu/pannotia/fw/Floyd-Warshall.cpp:
```diff
@@ -84,67 +91,150 @@
+	default:
+	    fprintf(stderr, "Unrecognized switch: -%c\n", opt);
+	    exit(1);
+	break;
```
This `break;` is unreachable.

mattsinc replied:
good point, @jebraun3 please fix this and submit a separate patch

BobbyRBruce commented on src/gpu/pannotia/fw/Floyd-Warshall.cpp:
```diff
@@ -84,67 +91,150 @@
+        fprintf(stderr, "Execution mode not specified! Use -h for help\n");
+        exit(1);
+    } else if (create_mmap && verify_results) {
+        fprintf(stdout, "Ignoring error checking\n");
```
I think the user should get explanation as to why. It may also be more apt as an stderr out.

Also, I think `verify_result` should be set to `false` here to ensure we do stop error checking.

mattsinc replied:
@jebraun3 this would be another thing from your cherry-picked commit.  I think it's fine to update the print message, but also could just be a comment.

BobbyRBruce replied:
Marking convo as resolved. Suggested change not for this PR.

BobbyRBruce commented on src/gpu/pannotia/fw/Floyd-Warshall.cpp:
```diff
@@ -84,67 +91,150 @@
+    } else if (create_mmap && verify_results) {
+        fprintf(stdout, "Ignoring error checking\n");
+    } else if (use_mmap && tmpchar != NULL) {
+        fprintf(stdout, "Ignoring input file\n");
```
This needs more explanation. I'd also say it probably should be an stderr out.

BobbyRBruce replied:
Marking as resolved. not for this PR

BobbyRBruce commented on src/gpu/pannotia/fw/Floyd-Warshall.cpp:
```diff
@@ -84,67 +91,150 @@
+        dim = *((int *)mmap(NULL, 1 * sizeof(int), PROT_READ, MAP_PRIVATE, fd, offset));
+        
+        // Read distmatrix in
+        int *distmatrixmap = (int *)mmap(NULL, (dim * dim + 1) * sizeof(int), PROT_READ | PROT_WRITE, MAP_PRIVATE, fd, offset);
```
There's a lot of long lines in this file. While the gem5-resources doesn't enforce any standards, it'd good if lines could be <80 characters long where possible.

mattsinc replied:
The line in question here is one @jebraun3 added in the cherry picked patch.  However, if we start enforcing the 80 character lines for benchmarks, I worry about how we can easily go back and compare them with the versions we inherited from.  When I was full-time at AMD I did this for every benchmark AMD had internally, and it caused endless problems later when trying to compare with original, non-gem5 versions.

BobbyRBruce commented on src/gpu/pannotia/fw/Floyd-Warshall.cpp:
I started noting every case but soon realized i'm better off making a general statement: There's a lot of bad indentation in this file. Could you go through and fix it?

mattsinc replied:
I believe @jebraun3 left the indentation that was inherited from the original Pannotia code, instead of changing it.  We certainly can update, but since it was part of the cherry-picked patches, I feel it should probably be part of a different patch?

BobbyRBruce replied:
Sure, this can wait a bit. Just would be nice to clean this up at some point.

BobbyRBruce replied:
Marking as resolved. Not to do in this PR.

BobbyRBruce commented on src/gpu/pannotia/fw/Floyd-Warshall.cpp:
```diff
@@ -84,67 +91,150 @@
+            printf("mmap.bin created!\n");
+            return 0;
+        }
+        free(adjmatrix);
```
I realize this is never used after this point but it's just hanging here while all the other memory is freed at the end of the party. I'd just move this down to be freed with the rest of the dynamic memory.

mattsinc replied:
@jebraun3 this would be another thing for you

mattsinc replied:
@BobbyRBruce from talking with @jebraun3 , we can't move this one because the array is declared inside the same else statement -- unless we move that declaration outside the else too, moving the free is not an option.

BobbyRBruce replied:
Ah, i see. To keep things neat I believe you can do this:

```cpp
// Declare dynamic memory variables at the start of code-block.
int* distmatrix = nullptr;
int* adjmatrix = nullptr;

// Do whatever you want here.
// ...

// At the end of codeblock.
// (Note: `free(nullptr)` is fine, no action occurs).
free(distmatrix)
free(adjmatrix)
```

Though, this is rather moot as we're freeing memory immediately before ceasing this process's execution at which point this memory will be freed anyway. I just think it's a good habit to have every dynamic memory allocation come with a deallocation.

BobbyRBruce commented on src/gpu/pannotia/fw/README.md:
````diff
@@ -19,10 +19,10 @@ To compile FW:
 ```
 
-If you use the Makefile.default file instead, the Makefile will generate code designed to run on the real GPU instead.  Moreover, note that Makefile.gem5-fusion requires you to set the GEM5_ROOT variable (either on the command line or by modifying the Makefile), because the Pannotia applications have been updated to use [m5ops](https://www.gem5.org/documentation/general_docs/m5ops/).  By default, ...
+If you use the Makefile.default file instead, the Makefile will generate code designed to run on the real GPU instead.  Moreover, note that Makefile.gem5-fusion requires you to set the GEM5_ROOT variable (either on the command line or by modifying the Makefile), because the Pannotia applications have been updated to use [m5ops](https://www.gem5.org/documentation/general_docs/m5ops/).  By default, ...
````
When writing in Markdown get into the habit of breaking your paragraphs into lines. This avoids massive lines of text like this. In some editors this will be very hard to read and deal with. There are generally two strategies:

1. Manually line-wrap given a certain character count per line.
2. Have each sentence be on its own line.

BobbyRBruce commented on src/gpu/pannotia/fw/README.md:
````diff
@@ -37,11 +37,31 @@ docker run -u $UID:$GID --volume $(pwd):$(pwd) -w $(pwd) gcr.io/gem5-test/gcn-gp
+
+```
+wget http://dist.gem5.org/dist/develop/datasets/pannotia/bc/1k_128k.gr
+docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID gcr.io/gem5-test/gcn-gpu gem5/build/GCN3_X86/gem5.opt gem5/configs/example/apu_se.py -n3 --mem-size=8GB --benchmark-root=gem5-resources/src/gpu/pannotia/fw/bin -c fw_hip.gem5 --options="-f 1k_128k.gr -m default"
````
I'd add `:latest` on the end of `gcn-gpu` just to make sure if the docker pulls the image we'll be pulling the latest version of it.

BobbyRBruce commented on src/gpu/pannotia/fw/README.md:
````diff
@@ -37,11 +37,31 @@ docker run -u $UID:$GID --volume $(pwd):$(pwd) -w $(pwd) gcr.io/gem5-test/gcn-gp
+To run FW using an mmapped input file, you must generate it first. An input file can be reused until it is overwritten by another file generation.  
+
+```
+docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID gcr.io/gem5-test/gcn-gpu gem5/build/GCN3_X86/gem5.opt gem5/configs/example/apu_se.py -n3 --mem-size=8GB --benchmark-root=gem5-resources/src/gpu/pannotia/fw/bin -c fw_hip.gem5 --options="-f 1k_128k.gr -m usemmap"
````
Same again here with the `:latest` tag.

BobbyRBruce commented on src/gpu/pannotia/sssp/README.md:
```diff
@@ -13,47 +13,55 @@ Single-Source Shortest Path (sssp) is a graph analytics application that is part
-git clone https://gem5.googlesource.com/public/gem5
-cd gem5
-docker run -u $UID:$GID --volume $(pwd):$(pwd) -w $(pwd) ghcr.io/gem5/gcn-gpu:latest scons build/GCN3_X86/gem5.opt -j <num cores>
+docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID ghcr.io/gem5/gcn-gpu bash -c "cd gem5-resources/src/gpu/pannotia/sssp ; export VARIANT=ELL ; make gem5-fusion"
```
I'd much prefer if you add `:latest` tags to your docker image names. This ensures if docker needs to pull it's getting the latest one. 

I won't mention this everywhere I see it, but there are other docker commands whichn require this.

BobbyRBruce commented on src/gpu/pannotia/pagerank/README.md:
````diff
@@ -13,47 +13,56 @@ PageRank (PR) is a graph analytics application that is part of the Pannotia benc
+docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID ghcr.io/gem5/gcn-gpu bash -c "cd gem5/util/m5 ; scons build/x86/out/m5"
 ```
+Note that Makefile.gem5-fusion requires you to set the GEM5_ROOT variable (either on the command line or by modifying the Makefile)  
+  
````
It's probably too late now but you basically say this 6 times across 6 README's. You could have a separate README in `src/gpu/pannotia` just for explaining how to build m5 and reference it as needed.

BobbyRBruce commented on src/gpu/pannotia/bc/README.md:
```diff
@@ -13,17 +13,8 @@ Betweenness Centrality (BC) is a graph analytics application that is part of the
-
-If you use the Makefile.default file instead, the Makefile will generate code designed to run on the real GPU instead.  Moreover, note that Makefile.gem5-fusion requires you to set the GEM5_ROOT variable (either on the command line or by modifying the Makefile), because the Pannotia applications have been updated to use [m5ops](https://www.gem5.org/documentation/general_docs/m5ops/).  By default, ...
-
+# Compiling BC and GCN3_X86/gem5.opt
```
I don't think the compilation of gem5 and BC need unified in one section. This header just makes the following two subheaders feel like we're repeating ourselves.

This seems to be a common pattern here so i'd advise fixing it in all cases.

BobbyRBruce commented on src/gpu/pannotia/bc/README.md:
````diff
@@ -34,9 +25,25 @@ cd gem5
 
-## Running BC on GCN3_X86/gem5.opt
+## Compiling BC
+The Pannotia applications have been updated to use [m5ops](https://www.gem5.org/documentation/general_docs/m5ops/). The docker command needs visibility to the gem5 repository for compilation and usage of the m5ops. Thus we run the docker command from a directory with visibility and cd into the folder before running the make/scons command.  
+  
+To compile m5ops assuming the gem5 repository is in your working directory:
+```
+docker run --rm -v ${PWD}:${PWD} -w ${PWD} -u $UID:$GID ghcr.io/gem5/gcn-gpu bash -c "cd gem5/util/m5 ; scons build/x86/out/m5"
````
Why is it we need to compile m5 in the docker container? I feel like we don't need to. I think we could just compile this from our own systems. If i'm wrong in this I think an explanation should be given in this text explaining why it is necessary.

BobbyRBruce commented on src/gpu/pannotia/pagerank/README.md:
````diff
@@ -13,47 +13,56 @@ PageRank (PR) is a graph analytics application that is part of the Pannotia benc
+# Running PageRank on GCN3_X86/gem5.opt
 
 The following command shows how to run the PageRank default version:
 
-# Assuming gem5 and gem5-resources are in your working directory
+## Assuming gem5 and gem5-resources are in your working directory
 ```
````
"The following..." doesn't work well being immediately followed by a header.


You do this in other instances, so apply the same where this exists in other files.

BobbyRBruce commented on src/gpu/pannotia/pagerank/README.md:
````diff
@@ -13,47 +13,56 @@ PageRank (PR) is a graph analytics application that is part of the Pannotia benc
 
-# Assuming gem5, pannotia (input graphs, see below), and gem5-resources are in your working directory
+## Assuming gem5, pannotia (input graphs, see below), and gem5-resources are in your working directory
 ```
````
Same here.

BobbyRBruce commented on src/gpu/pannotia/fw/Floyd-Warshall.cpp:
```diff
@@ -84,67 +91,150 @@
+    }    
+    
+    // Initialize the result matrix
+    result = (int *)malloc(dim * dim * sizeof(int));
+    if (!result) fprintf(stderr, "malloc failed - result\n");
```
Why don't we exit here? If malloc failes then we can't complete the execution, right?

mattsinc replied:
yes, we should -- @jebraun3 this is for you too

BobbyRBruce reviewed (changes requested):
I think there's a lot of redundancy in the individual GPU application READMEs. The build commands and requirement don't vary that much. If i were to do this again i'd advise maybe trying to have one unified README to save yourself all the copy and pasting. However, it's done now and it doesn't any harm. I just realize it must have been a lot to do.

KaiBatley committed:
resources: updated pannotia documentation for compile commands

KaiBatley commented:
@BobbyRBruce some of your comments are on the cherry picked commits already merged into develop. Mainly the Floyd-Warshall.cpp file from @jebraun3 which I have limited knowledge of. Probably best for @jebraun3 to handle those? I've pushed changes for the other comments and marked those resolved.

mattsinc commented:
And thanks/nice work @KaiBatley !

> I think there's a lot of redundancy in the individual GPU application READMEs. The build commands and requirement don't vary that much. If i were to do this again i'd advise maybe trying to have one unified README to save yourself all the copy and pasting. However, it's done now and it doesn't any harm. I just realize it must have been a lot to do.

I thought about this a lot when creating the original READMEs, but the way I understand how the website shows the benchmarks, it has a separate README for each benchmark.  Since that is what users see when they go to a specific benchmark, I did not see a good way to make it hierarchical in the way you are describing.  Happy to hear how to do this without requiring users to get information from multiple pages though.

BobbyRBruce commented:
> @BobbyRBruce some of your comments are on the cherry picked commits already merged into develop. Mainly the Floyd-Warshall.cpp file from @jebraun3 which I have limited knowledge of. Probably best for @jebraun3 to handle those? I've pushed changes for the other comments and marked those resolved.

Sure, that's understandable. There's definitely some funny code in that Floyd-Warshall.cpp, but given it's not directly related to this patch, my notes should be looked over and applied as different PR.

Outside of this, I have no issues with this PR.

BobbyRBruce commented:
Thumbs up from me. I'll leave it to one of you to merge this.

mattsinc commented:
Congrats @KaiBatley on your first merged commit!

@jebraun3 let's discuss @BobbyRBruce 's requests when we meet this week.
